### ISSUE TITLE
`jsdoc/check-param-names` should be fixable if order is wrong

### ISSUE THREAD

**Issue by ProLoser** [gajus/eslint-plugin-jsdoc#47]
I saw that another eslint plugin detects merely incorrect parameter order. It would be nice if this could be automatically fixed. You would have to ensure object properties are kept with param, and it could also (if this is preferred) move the `@returns` to the end.

**Comment by ProLoser** [comment]
Also useful for path references

**Comment by gajus** [comment]
PR is welcome.

**Comment by brettz9** [comment]
Duplicates could also be removed, extra items removed, and even bad names fixed, though any of these, especially the latter, could also be due to intending different params. Does anyone think any of these other cases should be fixable as well?

**Comment by golopot** [comment]
If the function param in jsdoc is inconsistent to that in code, which one is considered truth? Which one should be fixed?

**Comment by ProLoser** [comment]
There comments should be fixed, better to have documentation be updated rather than refactoring code.

**Comment by brettz9** [comment]
I'd think generally it'd be the code that was seen as truth, but yeah, that's the problem, it could be that the code isn't implemented as intended... And even for the cases of duplicates, a duplicate might indicate another param was intended, as with an extra param... But as @ProLoser suggests, there is a case for making it fixable in favor of the existing code too...

**Comment by brettz9** [comment]
While I think it would be better if ESLint had some way of passing in data from the command line to be used specifically for fixers (or formatters), we can leverage the fact that one can supply a specific rule with options at the command line, and create an option which tweaks the fixing behavior as desired. The option could be stored within one's config, or it could exclusively be used when one calls `eslint --fix`, e.g., `eslint --fix --rule='jsdoc/check-param-names': [2, {fix: ["badParamOrder", "duplicateParams", "extraParams", "badParamNames"]}]`, perhaps with our own CLI to simplify a little, e.g., `jsdoc-fix check-param-names -badParamOrder -duplicateParams -extraParams -badParamNames`

**Comment by golopot** [comment]
Fixing jsdoc makes more sense. Fixing code is too much work because the fixer needs to find and change all references of the param, and deal with name collision and shadowing.

**Comment by brettz9** [comment]
Yeah, I think so. It could be added later as an option (along the lines I was describing) if someone cared to implement, as an alternative fix to the jsdoc approach. We only need to settle on one as the default.

**Comment by brettz9** [comment]
But the other cases are questionable as to applying by default, so while some might operate by default, others should be opt in.

Here's what I'm thinking for default fixing:

- badParamOrder - on
- duplicateParams - on
- extraParams - off
- badParamNames - off

**Comment by morgan-coded** [comment]
Offering to take this one, fixer included. Two confirmations before I build:

1. Does the 2019 default set stand — `badParamOrder` on, `duplicateParams` on, `extraParams` off, `badParamNames` off?
2. For the order fixer: reorder whole tag blocks (descriptions riding along) to match the signature, with destructured/nested params left report-only in the first pass — or would you rather it land as a suggestion instead of a fix?

Happy to build to whichever shape you prefer.

**Comment by brettz9** [comment]
> Offering to take this one, fixer included.

Great!

> Two confirmations before I build:
> 
>     1. Does the 2019 default set stand — `badParamOrder` on, `duplicateParams` on, `extraParams` off, `badParamNames` off?

Yes, I think that is reasonable to keep.

>     2. For the order fixer: reorder whole tag blocks (descriptions riding along) to match the signature, with destructured/nested params left report-only in the first pass — or would you rather it land as a suggestion instead of a fix?

Unless you can think of specific risks for swapping parameters (beyond the risk that the developer intended the API to follow the docs rather than the implementation), I think it would be ideal to just automatically fix these to follow the implementation—yes, with descriptions riding along (bearing in mind that any text after the last tag—even after newlines—is still part of that last tag; there is no separate ending portion for JSDoc blocks). However, I would like to make `badParamOrder` and `duplicateParams` disable-able.

Perhaps `extraParams` and `badParamNames` could instead be offered as suggestions since dropping params or changing param names could be dangerous but sometimes useful.
